src/lib/functions/runtimes/js/builders/zisi.js

```diff
@@ -22,190 +17,45 @@ const addFunctionsConfigDefaults = (config) => ({
-    // We first check to see if the file being added is associated with any
-    // functions (e.g. restoring a file that has been previously deleted).
-    // If that's the case, we bundle just those functions.
-    const functionsWithPath = [...fileTree.entries()].filter(([, { inputs }]) => inputs.has(updatedPath))
-
-    if (functionsWithPath.length !== 0) {
-      await zipFunctionsAndUpdateTree({
-        fileTree,
-        functions: functionsWithPath.map(([, func]) => func),
-        sourceDirectory,
-        targetDirectory,
-        zipCache,
-        zipOptions,
-      })
-
-      return functionsWithPath.map(([mainFile]) => mainFile)
-    }
-
-    // We then check whether the newly-added file is itself a function. If so,
-    // we bundle it.
-    const functions = await getFunctions(sourceDirectory)
-    const matchingFunction = functions.find(({ mainFile }) => mainFile === updatedPath)
-
-    if (matchingFunction !== undefined) {
-      await zipFunctionsAndUpdateTree({
-        fileTree,
-        functions: [matchingFunction],
-        sourceDirectory,
-        targetDirectory,
-        zipCache,
-        zipOptions,
-      })
-
-      return [updatedPath]
-    }
-
-    // At this point, the newly-added file is neither a function nor a file
-    // associated with a function, so we can discard the update.
-    return
-  }
-
-  if (eventType === 'change' || eventType === 'unlink') {
-    // If the file matches a function's main file, we just need to operate on
-    // that one function.
-    if (fileTree.has(updatedPath)) {
-      const matchingFunction = fileTree.get(updatedPath)
-
-      // We bundle the function if this is a `change` event, or delete it if
-      // the event is `unlink`.
-      if (eventType === 'change') {
-        await zipFunctionsAndUpdateTree({
-          fileTree,
-          functions: [matchingFunction],
-          sourceDirectory,
-          targetDirectory,
-          zipCache,
-          zipOptions,
-        })
-      } else {
-        const { path: functionPath } = matchingFunction
-
-        fileTree.delete(updatedPath)
-
-        await del(functionPath, { force: true })
-      }
-
-      return [updatedPath]
-    }
-
-    // The update is in one of the supporting files. We bundle every function
-    // that uses it.
-    const functions = [...fileTree.entries()].filter(([, { inputs }]) => inputs.has(updatedPath))
+  return { buildPath, srcFiles }
+}
 
-    await zipFunctionsAndUpdateTree({
-      fileTree,
-      functions: functions.map(([, func]) => func),
-      sourceDirectory,
-      targetDirectory,
-      zipCache,
-      zipOptions,
+// Clears the cache for any files inside the directory from which functions are
+// served, except for any node_modules directories that live there.
+const clearFunctionsCache = (functionsPath) => {
+  Object.keys(require.cache)
+    .filter((key) => key.startsWith(functionsPath) && !functionsPath.includes(`${path.sep}node_modules${path.sep}`))
```

Comment:
I'm not clearing the cache for `node_modules` because that might lead to problems like https://github.com/netlify/cli/issues/2641. This does mean that if you install a new version of a module, you'll still have the old version in the `require.cache`.

However, I feel like this is a bit of an edge case, because it'd only happen if you opened a new terminal to `npm install` the new version without stopping the CLI process. If you're worried about this use case, perhaps we can try to think of an additional detection mechanism and show a warning message advising people to restart the CLI after a dependency update.

What do you think?

Reply:
We fixed that issue by using an [external dependency](https://github.com/netlify/cli/pull/2681).

`decache` handles clearing only the relevant modules and it is highly used:
https://www.npmjs.com/package/decache

So we would benefit from any possible potential bug fixes related to it.
Also the logic seems a bit different so it might be handling more use cases.

Reply:
I changed the clear cache function to use `decache`: https://github.com/netlify/cli/pull/2826/commits/f652a32aa859f1c697f70f5a69b3fc478393ecd9.